pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -517,13 +516,16 @@ private OpBatchReceive(CompletableFuture<Messages<T>> future) {
     }
 
     protected void notifyPendingBatchReceivedCallBack() {
-        final OpBatchReceive<T> opBatchReceive = pendingBatchReceives.poll();
+        final OpBatchReceive<T> opBatchReceive = getOpBatchReceive();
         if (opBatchReceive == null || opBatchReceive.future == null) {
             return;
         }
         notifyPendingBatchReceivedCallBack(opBatchReceive);
     }
 
+    private synchronized OpBatchReceive<T> getOpBatchReceive() {
```

Comment:
isn't pendingBatchReceives already an `ConcurrentLinkedQueue` ?

Reply:
Very true! static analyzer I used neither me realized that! Just rollbacked this :+1: